Dotty enum parsing

samuelchassot opened the pull request:
This add new structure to the AST to support Dotty Enum and add methods to the Parser to parse them.

samuelchassot committed:
Add AST structure and parsing methods to parse Dotty Enum

This commit add new AST structures to represent Enum, and Enum.Case. It adds methods in ScalametaParser.scala to parse Enum too.

allanrenucci commented:
The PR should target the `dotty-enum` branch

samuelchassot commented:
Oops I changed that

allanrenucci commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -2783,6 +2786,8 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
         funDefOrDclOrSecondaryCtor(mods)
       case KwType() =>
         typeDefOrDcl(mods)
+      case KwCase() if ahead(token.isNot[KwClass]) && ahead(token.isNot[KwObject])=>
```
```scala
case KwCase() if ahead(token.isNot[KwClass] && token.isNot[KwObject]) =>
```

allanrenucci replied:
Changes above need to take the dialect into account. I.e `dialect.allowEnums`

allanrenucci commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -2888,6 +2893,39 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
 
+  def caseDef(mods : List[Mod]) : Stat = {
+    accept[KwCase]
+    adjustSepRegions(Token.RightArrow(input, dialect, 0, 0))
```
Add the comment we have in Dotty:
```scala
// Scanner thinks it is in a pattern match after seeing the `case`.
// We need to get it out of that mode by telling it we are past the `=>`
```

allanrenucci commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -2888,6 +2893,39 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
+    }
+  }
+
+  def readRepeatedCase(mods : List[Mod]) : Defn = {
```
Can you use `def commaSeparated` here?

allanrenucci commented on scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala:
```diff
@@ -299,6 +299,49 @@ object Defn {
+      //          val p = parent.get
+      //          p.is[Template] && (p.parent.isEmpty || p.parent.get.is[Defn.Enum])
+      //        }
+      //      )
```
Can be removed

allanrenucci commented on scalameta/trees/shared/src/main/scala/scala/meta/Trees.scala:
```diff
@@ -299,6 +299,49 @@ object Defn {
+      //          val p = parent.get
+      //          p.is[Template] && (p.parent.isEmpty || p.parent.get.is[Defn.Enum])
+      //        }
+      //      )
```
Can be removed

allanrenucci commented on tests/shared/src/test/scala/scala/meta/tests/DottyEnumTests.scala:
```diff
@@ -0,0 +1,192 @@
+      |}
+    """.stripMargin,
+    
+    """Source(List(Defn.Enum(Nil, Type.Name("Foo"), Nil, Ctor.Primary(Nil, Name(""), Nil), """+
```
I think using multi-line strings across this file would be more readable. 
Here is an example of how this could be achieved.
https://github.com/scalameta/trees/blob/527cfabf82fad0e55d6af189517b822656735039/tests/shared/src/test/scala/scala/meta/tests/parsers/SbtSuite.scala#L57-L81

Try preserving the structure as well, like in this example

allanrenucci commented on tests/shared/src/test/scala/scala/meta/tests/DottyEnumTests.scala:
```diff
@@ -0,0 +1,192 @@
+    }
+  }
+
+  def checkBad(source : String) : Unit = {
```
I think `checkError` is a better name

allanrenucci commented on tests/shared/src/test/scala/scala/meta/tests/DottyEnumTests.scala:
```diff
@@ -0,0 +1,192 @@
+
+class DottyEnumTests extends FunSuite{
+  def checkOK(source : String, expected : String) : Unit = {
+    val name = source.slice(0, 50).replaceAll("\n", " ")
```
`slice(0, 50)` ~ `take(50)`

allanrenucci replied:
I think you can use `replace` instead of `replaceAll`

allanrenucci commented on tests/shared/src/test/scala/scala/meta/tests/DottyEnumTests.scala:
```diff
@@ -0,0 +1,192 @@
+    }
+  }
+
+  def checkOKParse(source : String) : Unit = {
```
I would name it checkOk as well. Having an overload is fine

allanrenucci commented on tests/shared/src/test/scala/scala/meta/tests/DottyEnumTests.scala:
```diff
@@ -0,0 +1,192 @@
+      val obtainedTree = Dotty(source).parse[Source]
+      assert(!obtainedTree.isInstanceOf[Parsed.Error])     
+    }
+  }
```
Some duplications that could be factored in something like:

```scala
def newParserTest(source: String)(body: Parsed[Source] => Unit) = {
  val name = source.take(50).replace('\n', ' ')
  test(name) {
    val parsed = Dotty(source).parse[Source]
    body(parsed)
  }
}

def checkOK(source: String, expected: String): Unit =
  newParserTest(source) { parsed =>
    assert(parsed.get.structure == expected)
  }

def checkOK(source: String): Unit = ...
def checkError(source: String, expected: String): Unit = ...
```

allanrenucci commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -543,15 +544,17 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
-      (token.is[KwCase] && token.isCaseClassOrObject)
+      (token.is[KwCase] && token.isCaseClassOrObject) ||
+      (token.is[KwCase] && ! token.isCaseClassOrObject) ||
+      token.is[KwEnum]
```
Changes above need to take the dialect into account. I.e `dialect.allowEnums`

allanrenucci commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -543,15 +544,17 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
     def unapply(token: Token): Boolean = {
       token.is[Modifier] || token.is[At] ||
-      token.is[KwClass] || token.is[KwObject] || token.is[KwTrait] ||
+      token.is[KwClass] || token.is[KwObject] || token.is[KwTrait] || token.is[KwEnum] ||
```
Changes above need to take the dialect into account. I.e `dialect.allowEnums`

allanrenucci commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -2905,6 +2943,8 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
       case KwObject() =>
         objectDef(mods)
+      case KwEnum() =>
+        enumDef(mods)
```
Changes above need to take the dialect into account. I.e `dialect.allowEnums`

allanrenucci commented on tests/shared/src/test/scala/scala/meta/tests/DottyEnumTests.scala:
```diff
@@ -0,0 +1,192 @@
+import scala.meta._
+import scala.meta.dialects.Dotty
+
+class DottyEnumTests extends FunSuite{
```
Follow project naming convention: `DottyEnumSuite`

allanrenucci commented on tests/shared/src/test/scala/scala/meta/tests/DottyEnumTests.scala:
```diff
@@ -0,0 +1,192 @@
+package scala.meta.tests
```
File should be moved to `tests/parser/`

allanrenucci commented on tests/shared/src/test/scala/scala/meta/tests/DottyEnumTests.scala:
```diff
@@ -0,0 +1,192 @@
+      |}
+      """.stripMargin
+    )   
+}
```
I think we need more test case. Some ideas:
- not ok: `enum Foo { case Bar { def hello = 1 } }`
- ok:
  ```scala
  enum Foo {
    case X,
         Y,
         Z
  }
  ```
- ok: `enum Bar extends X with Y with Z {}`

Try to think of more invalid test cases. For example where it could interact badly with existing feature (e.g. cases in pattern match)

allanrenucci reviewed (changes requested):
Formatting need some love. Things that I have seen in this PR:
- `x : Int` no space after `x`
- `identifier{` missing space before `{`
- `}else` missing space before `else`
- weird indentation
- redundant new lines before closing `}`. E.g:
  ```scala
  def foo = {
    ...
    bar

  }
  ```

allanrenucci commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -2783,6 +2787,8 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
         funDefOrDclOrSecondaryCtor(mods)
       case KwType() =>
         typeDefOrDcl(mods)
+      case KwCase() if ahead(dialect.allowEnums && token.isNot[KwClass] && token.isNot[KwObject]) =>
```
```scala
if dialect.allowEnums && ahead(token.isNot[KwClass] && token.isNot[KwObject])
```

allanrenucci commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -2905,6 +2941,8 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
         classDef(mods :+ atPos(casePos, casePos)(Mod.Case()))
       case KwObject() =>
         objectDef(mods)
+      case KwEnum() if (dialect.allowEnums) =>
```
parentheses not needed

allanrenucci commented on tests/shared/src/test/scala/scala/meta/tests/prettyprinters/DottyEnumSyntacticSuite.scala:
```diff
@@ -0,0 +1,59 @@
+    test(name) {
+      Dotty(source).parse[Source] match {
+        case Parsed.Success(t) =>
+          println(t.transform { case tree: Tree => tree.withOrigin(Origin.None) }.syntax)
```
Remove

allanrenucci commented on tests/shared/src/test/scala/scala/meta/tests/prettyprinters/DottyEnumSyntacticSuite.scala:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  testOK(
+    """enum Foo { case A[T](x: Int) extends Foo }"""
```
Only use single quote when triple quotes is not needed. This applies to the tests below as well

allanrenucci commented on tests/shared/src/test/scala/scala/meta/tests/prettyprinters/DottyEnumSyntacticSuite.scala:
```diff
@@ -0,0 +1,59 @@
+
+  testOK(
+    """enum Foo { class Bar }"""
+  )
```
Can you add a test for:
```scala
enum Foo {
  case Bar, Bat
  case Baz
}
```

allanrenucci commented on tests/shared/src/test/scala/scala/meta/tests/parsers/DottyEnumSuite.scala:
```diff
@@ -0,0 +1,517 @@
+            | Template(
+              |Nil,
+              | Nil,
+              | Self(Name(""), None), Nil))))""".trim.stripMargin.split('\n').mkString("")
```
I don't think you need to `trim`. Also I believe `split('\n').mkString("")` is equivalent to `.replace("\n", "")`. Comment applies as well to tests below

allanrenucci commented on scalameta/trees/shared/src/main/scala/scala/meta/internal/trees/ParentChecks.scala:
```diff
@@ -102,4 +102,8 @@ object ParentChecks {
   }
+
+  def CaseEnum(tree: Defn, parent: Tree, destination: String) : Boolean= {
+    parent.is[Template] && (parent.parent.isEmpty || parent.parent.get.is[Defn.Enum])
```
Let's comment the part that is always true:
```scala
parent.is[Template]
// ideally `&& (parent.parent.isEmpty || parent.parent.get.is[Defn.Enum])` but `parent.parent` is always empty
```

allanrenucci commented on scalameta/trees/shared/src/main/scala/scala/meta/internal/trees/ParentChecks.scala:
```diff
@@ -102,4 +102,8 @@ object ParentChecks {
     parent.is[Type]
   }
+
+  def CaseEnum(tree: Defn, parent: Tree, destination: String) : Boolean= {
```
format

allanrenucci commented on scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala:
```diff
@@ -594,6 +600,33 @@ object TreeSyntax {
           }
           s("case ", ppat, pcond, " ", kw("=>"), pbody)
 
+        // Enum.Case
```
Remove

allanrenucci commented on scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala:
```diff
@@ -594,6 +600,33 @@ object TreeSyntax {
+            throw new UnsupportedOperationException(s"$dialect doesn't support Enum")
+          }
+
+        //Enum.RepeatedCase
```
Remove

allanrenucci commented on scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala:
```diff
@@ -594,6 +600,33 @@ object TreeSyntax {
+            throw new UnsupportedOperationException(s"$dialect doesn't support Enum")
+          }
+
+        //Enum.Name
```
Remove

allanrenucci commented on scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala:
```diff
@@ -489,6 +489,12 @@ object TreeSyntax {
+            s(w(t.mods, " "), kw("enum"), " ", t.name, t.tparams, w(" ", t.ctor, t.ctor.mods.nonEmpty), templ(t.templ))
+          } else {
+            throw new UnsupportedOperationException(s"$dialect doesn't support Enum")
+          }
```
This is a bit verbose. Let's stick to the style used across the file:
```scala
if (!dialect. allowEnums) throw new UnsupportedOperationException(s"$dialect doesn't support enums")
s(...)
```
This comment applies to cases below as well

allanrenucci commented on scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala:
```diff
@@ -489,6 +489,12 @@ object TreeSyntax {
         case t: Defn.Var       => s(w(t.mods, " "), kw("var"), " ", r(t.pats, ", "), t.decltpe, " ", kw("="), " ", t.rhs.map(s(_)).getOrElse(s(kw("_"))))
         case t: Defn.Type      => s(w(t.mods, " "), kw("type"), " ", t.name, t.tparams, " ", kw("="), " ", t.body)
         case t: Defn.Class     => s(w(t.mods, " "), kw("class"), " ", t.name, t.tparams, w(" ", t.ctor, t.ctor.mods.nonEmpty), templ(t.templ))
+        case t: Defn.Enum     =>
```
Only one space before `=>`

allanrenucci commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -543,7 +544,9 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
       (token.is[Ellipsis] && token.next.is[DefIntro]) ||
-      (token.is[KwCase] && token.isCaseClassOrObject)
+      (token.is[KwCase] && token.isCaseClassOrObject) ||
+      (dialect.allowEnums && token.is[KwCase] && ! token.isCaseClassOrObject) ||
```
I think the 2 lines above are equivalent to:
```scala
(token.is[KwCase] && (token.isCaseClassOrObject || dialect.allowEnums))
```

allanrenucci commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -2888,6 +2894,36 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
+    }
+  }
+
+  def readCase(): Defn.Enum.Name = {
```
I think you should use the `def name` parser. I suggest adding a new `enumName` parser:
```scala
def enumName(): Term.Name = name(Enum.Name(_), advance = true)
```
Similar to:
https://github.com/scalameta/trees/blob/527cfabf82fad0e55d6af189517b822656735039/scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala#L1158-L1159

allanrenucci commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -2957,6 +2990,16 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
 
+  def enumDef(mods: List[Mod]): Defn.Enum = atPos(mods, auto) {
+    accept[KwEnum]
+    rejectMod[Mod.Override](mods, Messages.InvalidOverrideEnum)
```
This is partially broken in Dotty, but the only valid modifiers for enums are annotations and access modifier (`private`, `protected`). For enum cases, only annotations should be valid.

Can you add tests cases as well

samuelchassot replied:
Ok I'll change that. So an enum cannot be final ?

allanrenucci replied:
No, it is redundant. In `enum Foo { case Bar }`, you should never be able to extends `Foo` or `Bar`.

It is like `object`s. They are implicitly `final`. Although you can write `final object Foo` (it should probably be flagged as an error)

samuelchassot replied:
Oh I see, thank you !

allanrenucci commented on build.sbt:
```diff
@@ -1,10 +1,10 @@
+    // version ~= { dynVer =>
+    //   if (sys.env.contains("CI")) dynVer
+    //   else "SNAPSHOT" // only for local publishng
+    // },
```
Please. Don't commit this change

allanrenucci commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -467,6 +468,23 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
+      (implicit classifier: Classifier[Mod, M],
+      tag: ClassTag[M]) = {
+    mods.filter(m => !((mods.getAll[M]).contains(m))).foreach(m => syntaxError(errorMsg, at = m))
+  }
```
I am sure we can do better than this. What about:
```scala
def onlyAcceptMod(pred: Mod => Boolean)(mods: List[Mod], errorMsg: String) =
  mods.filterNot(pred).foreach(m => syntaxError(errorMsg, at = m))

// usage:
onlyAcceptMod(m => m.is[Mod.Private] || m.is[Mod.Protected] || m.is[Mod.Annot])(
  mods, "Only access modifiers are allowed on enum definitions")
```

allanrenucci commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -2888,6 +2912,31 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
+    // We need to get it out of that mode by telling it we are past the `=>`
+    adjustSepRegions(Token.RightArrow(input, dialect, 0, 0))
+
+    onlyAcceptMod[Mod.Annot](mods, "enum case can only have annotations")
```
I would say `Modifiers are not allowed on enum cases`

allanrenucci commented on scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:
```diff
@@ -2957,6 +3008,17 @@ class ScalametaParser(input: Input, dialect: Dialect) { parser =>
+    val enumName = typeName()
+    onlyAcceptMod[Mod.Private, Mod.Protected, Mod.Annot](
+      mods,
+      "enum can only have access modifier (private, protected) and annotation")
```
I would say: `Only access modifiers are allowed on enum definitions`

allanrenucci commented on tests/shared/src/test/scala/scala/meta/tests/parsers/DottyEnumSuite.scala:
```diff
@@ -0,0 +1,591 @@
+  checkError(
+    """
+      |enum Foo {
+      | public case A
```
There is no public modifier in Scala

allanrenucci commented on tests/shared/src/test/scala/scala/meta/tests/parsers/DottyEnumSuite.scala:
```diff
@@ -0,0 +1,583 @@
+import scala.meta._
+import scala.meta.dialects.Dotty
+
+class DottyEnumSuite extends FunSuite {
```
Let's reduce the verbosity of these tests. If a test can fit on a single line, write it on a single line. E.g.
```scala
// prefer
checkError("sealed enum Foo {}")

// to
checkError(
  "sealed enum Foo {}"
)
```
```scala
// prefer
checkError("enum Foo { lazy case A }")

// to
checkError(
  """
    |enum Foo {
    | lazy case A
    |}
  """.stripMargin
)
```

Group single line related test together. E.g.
```scala
checkError("enum Foo { final case A }")
checkError("enum Foo { lazy case A }")
```
This comment applies to this file and `DottyEnumSyntacticSuite.scala`

samuelchassot replied:
Ok I will change that. Do I put the source code on one line for the tests with expected tree too ?

allanrenucci replied:
As far as I can tell, no test with expected tree are one liner.

I am not saying all the tests should be one liner. There is a subset of tests such as the ones described above that read as well on a single one as on multiple lines.

Usually if a line goes above 80 to 100 characters, it should be split into multiple lines

samuelchassot replied:
I see. I made changes, tell me if it is fine now.
I put on one line the code if it was very small (e.g. "enum Foo { case A }") even if it was in a test with expected tree because one or two were already in this form.

allanrenucci reviewed (commented):
Oups! Forgot to submit a comment from my previous review

allanrenucci commented on scalameta/trees/shared/src/main/scala/scala/meta/internal/prettyprinters/TreeSyntax.scala:
```diff
@@ -594,6 +598,22 @@ object TreeSyntax {
+          else
+            s(kw("case"), " ", t.name, t.tparams, w(" ", t.ctor, t.ctor.mods.nonEmpty), " ", kw("extends"), " ", r(t.inits, " with "))
+
+
```
Remove extra blank line

allanrenucci commented on build.sbt:
```diff
@@ -1,10 +1,10 @@
+     version ~= { dynVer =>
+       if (sys.env.contains("CI")) dynVer
+       else "SNAPSHOT" // only for local publishng
+     },
```
Fix format here and the CI should be green

allanrenucci reviewed (approved):
LGTM. @olafurpg Do you want to have a look?

samuelchassot committed:
removes dialect from pretty printing

samuelchassot committed:
add a checkPos test

samuelchassot committed:
Remove checkPos test in DottyEnumSuite

samuelchassot committed:
Remove test with undesirable behaviour

Test in TreeToken that expected different tokens when called with different Dialect. That is no more expected.

samuelchassot committed:
Remove commented lines
